staging/src/k8s.io/client-go/rest/client.go

```diff
@@ -37,6 +37,15 @@ const (
 	envBackoffDuration = "KUBE_CLIENT_BACKOFF_DURATION"
 )
 
+const (
+	// default configuration for exponentially backoff delay with jitter
+	// with an initial delay of 250ms, maximum delay of 2m and
+	// a maximum of 10% jitter.
+	defaultBackoffInitialDelay = 250 * time.Millisecond
+	defaultBackoffMaxDelay     = 2 * time.Minute
```

Comment:
2m is a huge change compared to current "no-backoff".

I'm not saying that it's necessarily wrong one - we should just be careful with that.
It requires careful release note and I think a bit more soak-testing.

We should probably hold it until 1.24 and merge early in a cycle.

Reply:
yes, totally agree. I put it on hold with this goal in mind - we can merge it early 1.24 to give it enough soak time. I am not comfortable merging it this late for 1.23.

Reply:
Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500.  Is there any way in which this would be fine for 429, but somehow not fine for 50x?   E.g. I'm thinking about the way that maxRetries defaults to 10, so we could end up waiting for quite while if all 10 retries happen. That seems fine for 429s - since in that case the server is clearly alive and waiting is the right thing to do.  But what about when the statuses are 50x?  Is waiting for 10 retries with this level of backoff good?  I don't know... am just asking.

Reply:
> Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500.

I didn't yet carefully review this PR, but if that's true, I agree it's a problem.
We shouldn't be retrying 5xx (and IIRC we aren't), so backoff doesn't visibly affect those.

Reply:
I think we are retrying on 50x.  The code is this:

```
func (b *URLBackoff) UpdateBackoff(actualUrl *url.URL, err error, responseCode int) {
	// range for retry counts that we store is [0,13]
	if responseCode > maxResponseCode || serverIsOverloadedSet.Has(responseCode) {
		b.Backoff.Next(b.baseUrlKey(actualUrl), b.Backoff.Clock.Now())
		return
```
(here https://github.com/kubernetes/client-go/blob/ad95a77727e65f037cc877600f4d8ea4d867c617/rest/urlbackoff.go#L86) 

and maxResponseCode is defined to be 499.

Reply:
Makes me wonder if maybe this PR _is_ the answer afterall, with some form of dynamic adjustment as per the comments https://github.com/kubernetes/kubernetes/pull/97430

Reply:
> Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500

@wojtek-t If the server wants the caller to retry the request then it sends the following response to the caller:
```
StatusCode = {429|5xx}
Header:
  retry-After: N
```
A response status code of `503` alone from the server will not trigger a retry from the client. This is documented here:
https://github.com/kubernetes/kubernetes/blob/b429c4c9e1b3004953dfa42e46ac901bdc9e4460/staging/src/k8s.io/client-go/rest/with_retry.go#L178-L189

@JohnRusk  what you are pointing to is `UpdateBackoff` updating the backoff parameters. I know this is a little bit confusing. I am going to work on a clean up PR to make it less confusing.


> I'm thinking about the way that maxRetries defaults to 10, so we could end up waiting for quite while if all 10 retries happen

Yes, I agree, in the worst case the client can end up waiting for quite a while before giving up, especially when the request does not have a context deadline. I would like to let this PR soak for a while to see how adversely it affects.

Reply:
Ah, thank you @tkashem .  I had seen that it updated the backoff params, and forgotten that the params are only _used_ if there's a Rety-After.

BTW, I've been looking at ListWatch functionality in kublet. In that specific case, there's actually [an extra layer of retries](https://github.com/kubernetes/kubernetes/blob/b429c4c9e1b3004953dfa42e46ac901bdc9e4460/staging/src/k8s.io/client-go/tools/cache/reflector.go#L218) in addition to what we see in `Request.request()`.  That has the effect that the app does do backoff for 50x errors.  But at least its exponential by default.  I'm not sure whether we want to fix that for 500x or not. I guess the idea is that the kubelet will just keep trying and trying until the API server comes back online, and so I suppose the current behaviour there might be correct (and somewhat risky to change, due to risk of impacting kubelets ability to recover).  So I'm commenting on it here mostly just for completeness.

Reply:
Yes - Kubelet has an additional layer. But we should get rid of it.
Instead of adding this in the kubelet itself we should have fixed & improved error handingling in client-go.
With the changes that are happening in this area now (we may need to go a bit futher), we should be able to get rid of this custom logic in kubelet.

Reply:
@wojtek-t @JohnRusk does kubelet have an additional retry layer on top of the reflector, can you please point me to it?

Reply:
I actually counted about 7 different pieces of retry logic in Kubelet!   

Sorry about the slow reply, I've been on holiday and only just got back.  I'll try to share a list of them all here.

Reply:
Ah, it looks like you've seen some of them already @tkashem  , here https://github.com/kubernetes/kubernetes/pull/102107#discussion_r635449646.  The specific one I was talking about is as per the hyperlink in my earlier comment.

Reply:
FYI @tkashem and @wojtek-t , as far as I can tell, here's the complete list of all the retry logic in Kubelet.  I'm listing it here for completeness, to answer takashem's question, and to highlight the last two points to you @wojtek-t  - since they look potentially troublesome.  They're not configurable, and their inital backoff intervals are small.

Hyperlinks point to the source code.  

1. [HTTP internal retry](https://github.com/golang/net/blob/fe4d6282115f0528fd5bc04b5644ce8ba1f43a50/http2/transport.go#L501), only for selected errors (errClientConnUnusable, errClientConnGotGoAway, ErrCodeRefusedStream aka REFUSED_STREAM).  I wonder if we get errClientConnUnusable after certain other errors earlier on the connection, but apart from that I don't think this retry mechnanism is very significant to us.
2. [Request.go main backoff](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/rest/request.go#L970). This is the subject of this current PR.  It will retry all errors which either have a Retry-After sent from the server (typically 429s) OR which look like a connection reset or app server shutdown. It will never retry other errors (e.g. TLS handshake errors, which can be common on some highly loaded servers at startup). Such errors are handled only by the various hard-coded backoffs listed below, as items 4 through 7.
3. [Request.go also respects Retry-After](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/rest/request.go#L1014). In addition to the backoff value it calculates, Request.go will also sleep for the time specified in the Retry-After response header. In practice, kube apiserver always returns 1 second as the value in that header. (As discussed here https://github.com/kubernetes/kubernetes/issues/79805)
4. [ListWatch (connections)](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/tools/cache/reflector.go#L421). Hard-coded exponential backoff.  This was enhanced by PR #100878 to backoff on 429s.  That PR adds a layer of backoff for 429s over and above any done by the Request level backoff above. Although I'm not sure how necessary that will be going forward (i.e. if this current PR is approved).  Perhaps that fits with wojeck-t's comment above.
5. [ListWatch (outer loop)](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/tools/cache/reflector.go#L220). This is the one I mentioned earlier tkashem.  It's a hard-coded exponential backoff.
6. [csi-plugin waitForAPIServerForEver](https://github.com/kubernetes/kubernetes/blob/f0dbc32ed9c8a51a6d149eb62d868fb995f0db3a/pkg/volume/csi/csi_plugin.go#L1053).  This appears to be hard coded a 1 second delay,with no backoff.
7. [csi-pluging main backoffs](https://github.com/kubernetes/kubernetes/blob/f0dbc32ed9c8a51a6d149eb62d868fb995f0db3a/pkg/volume/csi/csi_plugin.go#L292).  Two levels of nested exponential backoff, with hard-coded parameters.  If all requests fail, it will retry a total of 24 times with a total waiting time of 150 seconds, then restart the kubelet.   The intial backoff interval is only 15 milliseconds.  Which doesn't sound very big, especially in the context of a large cluster with thousands of kubelets.

Reply:
Two more, both with no backoff (in most cases) and hard coded-retry limit.
9. kubelet_node_statsu.go: tryUpdateNodeStatus
10. k8s.io/component-helpers/apimachinery/lease/controller.go: retryUpdateLease.

I will probably post more about these two in a new Issue, but need to investigate them a little more first.

Reply:
The issue with those last two turns out to be not primarily a retry issue. They are _designed_ to reset all TCP connections after only the second failure.  But that causes a problem. In fact, it turns out to be the _only_ significant retry-related problem that I've found when testing with 5000 nodes.  Fixing that one problem is necessary and sufficient to solve all the apparent retry problems in my 5000 node test clusters.    I.e. when I solved that, issue #65954 was solved on my cluster.  But wiwthout that fix, nothing else could solve #65954.

I've logged a PR here to fix it #107781.